### PR TITLE
Fixed #34752 -- Fixed handling ASGI http.disconnect for streaming responses.

### PR DESCRIPTION
**Pull request by qxcv**
See this ticket for details: https://code.djangoproject.com/ticket/34752

This PR:

- [x] Adds some tests to ensures that ASGI `http.disconnect` is handled correctly during `StreamingHttpResponses`.
- [x] Modifies the existing ASGI handler so that the test passes.
- [x] Updates the `StreamingHttpResponse` documentation to document this new feature.

Additional context from a comment I made on the ticket:

> One subtlety: I originally made a smaller modification to the handler that put view logic and response logic in the same coroutine, but otherwise  followed the same flow. I found this broke in practice because sometimes  //both// coroutines would return at once. This caused the previous  `pending.pop()`/`done.pop()` logic to error out because one of those two  sets would be empty. I think that this problem is particularly likely to  happen in situations where some coroutines are long-lived. I haven't added  a test for this edge case yet, though.

### CODE REVIEW

**Review by carltongibson — COMMENTED** [review]
Just an initial comment; still looking 👀

File: django/core/handlers/asgi.py [hunk 1]
```diff
@@ -178,38 +178,44 @@ async def handle(self, scope, receive, send):
...
-        try:
-            response = done.result()
-        except RequestAborted:
-            body_file.close()
-            return
-        except AssertionError:
-            body_file.close()
-            raise
-        # Send the response.
+        with body_file:
```
Comment (carltongibson) [inline django/core/handlers/asgi.py:181]: This might be a worthwhile refactoring (or not) but it would be nice to have it as a separate change to not obscure the diff here.

File: django/core/handlers/asgi.py [hunk 2]
```diff
@@ -178,38 +178,44 @@ async def handle(self, scope, receive, send):
...
+            tasks = [
+                # We will check the status of these tasks and (optionally)
+                # terminate them in this order. The listen_for_disconnect task
+                # goes first because it should not raise unexpected errors that
+                # would prevent us from cancelling get_and_send_response.
+                asyncio.create_task(self.listen_for_disconnect(receive)),
+                asyncio.create_task(self.get_and_send_response(request, send)),
+            ]
+            await asyncio.wait(tasks, return_when=asyncio.FIRST_COMPLETED)
+            # Now wait on both tasks (they may have both finished by now).
```
Comment (carltongibson) [inline django/core/handlers/asgi.py:201]: @qxcv Can I ask what motivates this change? (Thanks!)

Comment (qxcv) [reply django/core/handlers/asgi.py:201]: Assuming you mean the change from

```
done, pending = await asyncio.wait(tasks, return_when=asyncio.FIRST_COMPLETED)
done, pending = done.pop(), pending.pop()
# some conditionals to handle done and pending
```

to

```
await asyncio.wait(tasks, return_when=asyncio.FIRST_COMPLETED)
# a loop through tasks
```

The answer is that both tasks can finish at the same time, which means that both tasks end up in `done` and the `pending.pop()` errors out. I only observed this behavior when serving an app with Daphne, not in the existing unit tests. I tried to write a test for it last week but my asyncio-foo wasn't up to the task of forcing both tasks to finish at once. I'll try again later this week (I found [this great overview of asyncio](https://tenthousandmeters.com/blog/python-behind-the-scenes-12-how-asyncawait-works-in-python/) that cleared up a lot of things for me).

Comment (carltongibson) [reply django/core/handlers/asgi.py:201]: > The answer is that both tasks can finish at the same time, which means that both tasks end up in done and the pending.pop() errors out. I only observed this behavior when serving an app with Daphne, not in the existing unit tests. I tried to write a test for it last week but my asyncio-foo wasn't up to the task of forcing both tasks to finish at once.

OK... that's really interesting 🤔 Since only a single task gets run on the event loop at a time, how they finish simultaneously is a curious… — but **yes** if that's happening that change make sense. 

If you can't come up with a test, maybe a minimal project that demonstrates would be good to see (if you can 🙏)

**Review by carltongibson — COMMENTED** [review]
OK, I think this makes sense.

**Comment by qxcv** [comment]
Heads up: I'm pretty busy with work at the moment and may not be able to finish this for a week or two. Going to set a reminder to myself to come back to this next weekend.

**Comment by qxcv** [comment]
Addressed the review comments above.

I also added a test that _sometimes_ causes both tasks in the task list to return at once (`test_disconnect_both_return`). You can run it with:

```python
cd tests; ./runtests.py -k disconnect_both_return
```

I added a hacky print to `ASGIHandler.handle()` which sometimes says `Number of finished tasks: 2`, which at least shows that it can sometimes happen in non-contrived circumstances. However, it's nondeterministic (perhaps the tests are running in a multithreaded executor?). Also, the test only succeeds at inducing this behavior as an accident of the way that `ASGICommunicator` works—if the implementation details changed slightly then it would break.

Short of artificially inserting a barrier at the end of the two tasks, I don't see a way to reliably cause both to return at once in a test. Given this, I'm inclined to rip out the extra test (and the `print`) and call it a day. Thoughts?

**Comment by qxcv** [comment]
Also I was curious how two tasks can finish at the same time in `wait()`, so I had a look at the source for `BaseEventLoop` in asyncio. Here's the hand-wavy explanation I put together for myself, working backwards from the event loop:

* The part of the event loop that actually runs tasks/callbacks is [in `_run_once()`](https://github.com/python/cpython/blob/4b4a5b70aa8d47b1e2a0582b741c31b786da762a/Lib/asyncio/base_events.py#L1935-L1951). It looks at all the tasks that are ready to run (a list named `self._ready`) and runs each of them turn. It then returns control back to a higher level of the event loop.
* The list of ready tasks in `self._ready` is [populated by `_call_soon(task)`](https://github.com/python/cpython/blob/4b4a5b70aa8d47b1e2a0582b741c31b786da762a/Lib/asyncio/base_events.py#L791-L796). All `_call_soon(task)` does is put the task in the `self._ready` list so that it will be executed at the next iteration of the event loop.
* `call_soon()` gets invoked in various places. The most important one is [`Future.__schedule_callbacks()`](https://github.com/python/cpython/blob/531930f47f6b2a548d31e62cb4ad3e215a24bf53/Lib/asyncio/futures.py#L161-L173), which is in turn executed every time a caller does `future.set_result()`, `future.exception()`, etc. etc. In other words: setting the result of a future causes anything that was waiting on that future to run at the next tick of the event loop.
* Now we can look at what `asyncio.wait()` does. The core functionality is in [`_wait()`](https://github.com/python/cpython/blob/main/Lib/asyncio/tasks.py#L500-L541), which:
  * Creates a new future called `waiter`.
  * Adds a callback to each of the tasks that runs when the corresponding task finishes. This callback records the number of tasks that have finished so far. If sufficiently many have finished (either one tasks or all tasks, depending on the value of `return_when`), then it calls `waiter.set_result()` (or `waiter.exception()`).
  * `_wait()` then awaits `waiter` future. Thus it will only get control back at the next tick after one of the tasks finishes.
* Putting it all together: the current code calls `asyncio.wait(tasks, return_when=asyncio.FIRST_COMPLETED)`. As soon as one of the tasks in the `tasks` list finishes, the callback inside `_wait` will schedule `waiter` to wake up at the next tick of the event loop, which in turn causes the `asyncio.wait()` to return. However, it's possible for both tasks to finish at the same tick of the event loop, in which case they will both be marked as `done()` by the time that `_wait()` gets control back from `await waiter` at the next tick of the loop.

File: tests/asgi/tests.py [hunk 3]
```diff
@@ -402,3 +426,95 @@ class TestASGIHandler(ASGIHandler):
...
+        communicator = ApplicationCommunicator(application, scope)
+        await communicator.send_input({"type": "http.request"})
+        await communicator.receive_output(timeout=1)
+        first_response = await communicator.receive_output(timeout=1)
+        self.assertEqual(first_response["body"], b"foo\n")
+        # Simulate disconnection.
+        await communicator.send_input({"type": "http.disconnect"})
+        # Ensure that 'bar\n' isn't sent.
+        with self.assertRaises(asyncio.TimeoutError):
+            await communicator.receive_output(timeout=0.2)  # type: ignore
```
Comment (felixxm) [inline tests/asgi/tests.py:520]: I can see `ResourceWarning` when running `asgi` tests on Python 3.12:
```
$ python -Wall ./runtests.py asgi.tests.ASGITest --parallel=1
Testing against Django installed in '/django/django'
Found 20 test(s).
System check identified no issues (0 silenced).
....................
----------------------------------------------------------------------
Ran 20 tests in 8.087s

OK
Exception ignored in: <function SpooledTemporaryFile.__del__ at 0x7f66ad4b8540>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/tempfile.py", line 736, in __del__
    _warnings.warn(
ResourceWarning: Unclosed file <tempfile.SpooledTemporaryFile object at 0x7f66ab688cd0>
```

File: tests/asgi/tests.py [hunk 4]
```diff
@@ -237,6 +237,30 @@ async def test_disconnect(self):
         with self.assertRaises(asyncio.TimeoutError):
             await communicator.receive_output()
 
+    async def test_disconnect_both_return(self):
```
Comment (felixxm) [inline tests/asgi/tests.py:240]: `test_disconnect_both_return()` and `test_streaming()` work without this patch. Extra test coverage is welcome but we should move it to a separate commit/PR.

File: tests/asgi/tests.py [hunk 5]
```diff
@@ -237,6 +237,30 @@ async def test_disconnect(self):
...
+        # This test forces both the disconnect listener and the task that sends
+        # the resposne to finish at the same time.
```
Comment (felixxm) [inline tests/asgi/tests.py:242]: Typo
```suggestion
        # This test forces both the disconnect listener and the task that sends
        # the response to finish at the same time.
```

File: tests/asgi/tests.py [hunk 6]
```diff
@@ -237,6 +237,30 @@ async def test_disconnect(self):
         with self.assertRaises(asyncio.TimeoutError):
             await communicator.receive_output()
 
+    async def test_disconnect_both_return(self):
+        # This test forces both the disconnect listener and the task that sends
```
Comment (felixxm) [inline tests/asgi/tests.py:241]: Try to omit prefixes like `This test`, `Ensure`, `Test`, etc.

File: docs/ref/request-response.txt [hunk 7]
```diff
@@ -1282,6 +1282,39 @@ Attributes
     This is useful for middleware needing to wrap
     :attr:`StreamingHttpResponse.streaming_content`.
 
+
```
Comment (felixxm) [inline docs/ref/request-response.txt:1285]:
```suggestion
```

File: docs/ref/request-response.txt [hunk 8]
```diff
@@ -1282,6 +1282,39 @@ Attributes
...
+        except asyncio.CancelledError:
+            # Clean up resources here and re-raise.
+            clean_up()
+            raise
+
+
+    async def my_streaming_view(request):
+        return StreamingHttpResponse(streaming_response())
+
+This example only shows how to handle client disconnection while the response is
```
Comment (felixxm) [inline docs/ref/request-response.txt:1312]: Wrap at 79 chars.

File: docs/topics/async.txt [hunk 9]
```diff
@@ -197,6 +197,10 @@ cleanup::
             # Handle disconnect
             raise
 
+You can also handle client disconnects in streaming responses; see
```
Comment (felixxm) [inline docs/topics/async.txt:200]: Try to omit `;` in docs.

File: django/core/handlers/asgi.py [hunk 10]
```diff
@@ -187,30 +187,40 @@ async def handle(self, scope, receive, send):
...
             return
+
+        async def process_request(request, send):
+            response = await self.run_get_response(request)
+            await self.send_response(response, send)
+
         # Try to catch a disconnect while getting response.
         tasks = [
-            asyncio.create_task(self.run_get_response(request)),
+            # We will check the status of these tasks and (optionally)
```
Comment (felixxm) [inline django/core/handlers/asgi.py:197]: Please omit `We` in comments
```suggestion
            # Check the status of these tasks and (optionally) ...
```

**Review by felixxm — COMMENTED** [review]
@qxcv Thanks :+1: I left initial comments.

File: tests/asgi/tests.py (hunk 3 above)
Comment (ngnpope) [reply tests/asgi/tests.py:520]: There is no typing in Django, so this suppression comment can be removed.

```suggestion
            await communicator.receive_output(timeout=0.2)
```

**Review by ngnpope — CHANGES_REQUESTED** [review]

**Comment by carltongibson** [comment]
Ref https://github.com/django/django/pull/17147#issuecomment-1687233828 — Good follow up @qxcv, thanks! 

> As soon as one of the tasks in the tasks list finishes, the callback inside _wait will schedule waiter to wake up at the next tick of the event loop, which in turn causes the asyncio.wait() to return. However, it's possible for both tasks to finish at the same tick of the event loop, in which case they will both be marked as done() by the time that _wait() gets control back from await waiter at the next tick of the loop.

👀

**Comment by felixxm** [comment]
@andrewgodwin Can you take a look? (it is a Django 5.0 release blocker)

File: django/core/handlers/asgi.py (hunk 2 above)
Comment (carltongibson) [reply django/core/handlers/asgi.py]: I think this is resolved by the discussion below https://github.com/django/django/pull/17147#issuecomment-1689393302 

If someone has a strategy to test it then super, but checking both tasks (in a loop) can't hurt... — it's just that the initial implementation **assumed** one task would be still pending.

**Comment by felixxm** [comment]
@qxcv Thanks :+1: Welcome aboard :boat: I pushed edits to fix comments.

There is only for remaining issue.

File: tests/asgi/tests.py [hunk 11]
```diff
@@ -254,7 +279,7 @@ async def test_assert_in_listen_for_disconnect(self):
         await communicator.send_input({"type": "http.not_a_real_message"})
         msg = "Invalid ASGI message after request body: http.not_a_real_message"
         with self.assertRaisesMessage(AssertionError, msg):
-            await communicator.receive_output()
+            await communicator.wait()
```
Comment (felixxm) [inline tests/asgi/tests.py:282]: This causes `ResourceWarning` on Python 3.12+:
```
./runtests.py asgi.tests.ASGITest.test_assert_in_listen_for_disconnect --parallel=1
Testing against Django installed in '/django/django'
Found 1 test(s).
System check identified no issues (0 silenced).
.
----------------------------------------------------------------------
Ran 1 test in 0.004s

OK
Exception ignored in: <function SpooledTemporaryFile.__del__ at 0x7ff588dc09a0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/tempfile.py", line 736, in __del__
    _warnings.warn(
ResourceWarning: Unclosed file <tempfile.SpooledTemporaryFile object at 0x7ff586f9da50>
```
Any ideas?

Comment (felixxm) [reply tests/asgi/tests.py:282]: I added a possible fix. Can anyone take a look?

Comment (carltongibson) [reply tests/asgi/tests.py:282]: @felixxm The fix in  08e72a6413bbd66e4d39a89b5b9c31c38408289f looks correct: it's restoring the behaviour from before this change. (Note the same `except` block in the previous version.) The test exercises the `assert` for the (protocol server programming error) invalid message type in `listen_for_disconnect()`. 

It would have been dropped here because @qxcv proposed moving to use a context manager block for the body file here. See https://github.com/django/django/pull/17147#pullrequestreview-1564929017

👍

**Review by carltongibson — APPROVED** [review]

File: tests/asgi/tests.py (hunk 11 above)
Comment (felixxm) [reply tests/asgi/tests.py:282]: Thanks for checking :+1: 

> That may be worth it, if you think, as an improvement — but a part of me thinks there **really should** only be a small finite set of ways out of this block and I'd rather blow up if we've missed one... 🤔

Yeah, we can leave it for a separate discussion/PR.
